data/transactions/logic/assembler.go

```diff
@@ -0,0 +1,1112 @@
+		dis.pendingLabels = make(map[int]string)
+	}
+	dis.pendingLabels[target] = label
+}
+
+type disassembleFunc func(dis *disassembleState)
+
+type disassembler struct {
+	name    string
+	handler disassembleFunc
+}
+
+var disassemblers = []disassembler{
+	{"intcblock", disIntcblock},
+	{"intc", disIntc},
+	{"bytecblock", disBytecblock},
+	{"bytec", disBytec},
+	{"arg", disArg},
+	{"txn", disTxn},
+	{"gtxn", disGtxn},
+	{"global", disGlobal},
+	{"bnz", disBnz},
+	{"load", disLoad},
+	{"store", disStore},
+}
+
+var disByName map[string]disassembler
+
+func init() {
+	disByName = make(map[string]disassembler)
+	for _, dis := range disassemblers {
+		disByName[dis.name] = dis
+	}
+}
+
+var errTooManyIntc = errors.New("intcblock with too many items")
+
+func parseIntcblock(program []byte, pc int) (intc []uint64, nextpc int, err error) {
+	pos := pc + 1
+	numInts, bytesUsed := binary.Uvarint(program[pos:])
+	if bytesUsed <= 0 {
+		err = fmt.Errorf("could not decode int const block size at pc=%d", pos)
+		return
+	}
+	pos += bytesUsed
+	if numInts > uint64(len(program)) {
+		err = errTooManyIntc
+		return
+	}
+	intc = make([]uint64, numInts)
+	for i := uint64(0); i < numInts; i++ {
+		if pos >= len(program) {
+			err = fmt.Errorf("bytecblock ran past end of program")
+			return
+		}
+		intc[i], bytesUsed = binary.Uvarint(program[pos:])
+		if bytesUsed <= 0 {
+			err = fmt.Errorf("could not decode int const[%d] at pc=%d", i, pos)
+			return
+		}
+		pos += bytesUsed
+	}
+	nextpc = pos
+	return
+}
+
+func checkIntConstBlock(cx *evalContext) int {
+	pos := cx.pc + 1
+	numInts, bytesUsed := binary.Uvarint(cx.program[pos:])
+	if bytesUsed <= 0 {
+		cx.err = fmt.Errorf("could not decode int const block size at pc=%d", pos)
+		return 1
+	}
+	pos += bytesUsed
+	if numInts > uint64(len(cx.program)) {
+		cx.err = errTooManyIntc
+		return 0
+	}
+	//intc = make([]uint64, numInts)
+	for i := uint64(0); i < numInts; i++ {
+		if pos >= len(cx.program) {
+			cx.err = fmt.Errorf("bytecblock ran past end of program")
+			return 0
+		}
+		_, bytesUsed = binary.Uvarint(cx.program[pos:])
+		if bytesUsed <= 0 {
+			cx.err = fmt.Errorf("could not decode int const[%d] at pc=%d", i, pos)
+			return 1
+		}
+		pos += bytesUsed
+	}
+	cx.nextpc = pos
+	return 1
+}
+
+var errShortBytecblock = errors.New("bytecblock ran past end of program")
+var errTooManyItems = errors.New("bytecblock with too many items")
+
+func parseBytecBlock(program []byte, pc int) (bytec [][]byte, nextpc int, err error) {
+	pos := pc + 1
+	numItems, bytesUsed := binary.Uvarint(program[pos:])
+	if bytesUsed <= 0 {
+		err = fmt.Errorf("could not decode []byte const block size at pc=%d", pos)
+		return
+	}
+	pos += bytesUsed
+	if numItems > uint64(len(program)) {
+		err = errTooManyItems
+		return
+	}
+	bytec = make([][]byte, numItems)
+	for i := uint64(0); i < numItems; i++ {
+		if pos >= len(program) {
+			err = errShortBytecblock
+			return
+		}
+		itemLen, bytesUsed := binary.Uvarint(program[pos:])
```

Comment:
`itemLen` can be negative when cast as `int`, which can cause an OOB slice below